File: Interview Problems/Parking Lot/Java Code/controller/ExitController.java
package controller;

import domain.Receipt;
import domain.Ticket;
import service.PaymentService;
import service.PricingService;
import service.ReceiptService;
import service.SlotService;
import service.TicketService;

import java.util.Optional;
import java.util.UUID;

public class ExitController {

    private final TicketService ticketService;
    private final PricingService pricingService;
    private final PaymentService paymentService;
    private final ReceiptService receiptService;
    private final SlotService slotService;

    public ExitController(TicketService ticketService,
                          PricingService pricingService,
                          PaymentService paymentService,
                          ReceiptService receiptService,
                          SlotService slotService) {
        this.ticketService = ticketService;
        this.pricingService = pricingService;
        this.paymentService = paymentService;
        this.receiptService = receiptService;
        this.slotService = slotService;
        System.out.println("[CONTROLLER] ExitController initialized");
    }

    public ExitResult exitVehicle(UUID ticketId) {
        System.out.println("[CONTROLLER] Vehicle exit request - Ticket: " + ticketId);

        try {
            // 1. Retrieve ticket
            Optional<Ticket> ticketOpt = ticketService.getTicket(ticketId);
            if (ticketOpt.isEmpty()) {
                return new ExitResult(false, null, 0.0, "Ticket not found");
            }

            Ticket ticket = ticketOpt.get();

            // 2. Guard: ticket must still be active
            if (!ticket.isActive()) {
                return new ExitResult(false, null, 0.0, "Ticket is not active");
            }

            // 3. Calculate fee
            double fee = pricingService.calculateFee(ticket);
            System.out.println("[CONTROLLER] Fee calculated: ₹" + fee);

            // 4. Process payment with retry (max 3 attempts)
            // Slot is intentionally NOT released until payment succeeds
            boolean paymentSuccess = paymentService.processPaymentWithRetry(ticketId, fee, 3);
            if (!paymentSuccess) {
                return new ExitResult(false, null, fee, "Payment failed after retries");
            }

            // 5. Generate receipt and mark as paid
            Receipt receipt = receiptService.generateReceipt(ticket, fee);
            receiptService.markReceiptAsPaid(receipt);

            // 6. Release slot — only after confirmed payment
            slotService.releaseSlot(ticket.getSlotId());

            // 7. Deactivate ticket to prevent reuse
            ticketService.deactivateTicket(ticketId);

            System.out.println("[CONTROLLER] Vehicle exit successful - Receipt: " + receipt.getId());
            return new ExitResult(true, receipt.getId(), fee, "Exit successful");

        } catch (Exception e) {
            System.out.println("[CONTROLLER] Vehicle exit failed: " + e.getMessage());
            return new ExitResult(false, null, 0.0, e.getMessage());
        }
    }

    public String generateReceiptText(UUID ticketId) {
        Optional<Ticket> ticketOpt = ticketService.getTicket(ticketId);
        if (ticketOpt.isEmpty()) return "Ticket not found: " + ticketId;

        Ticket ticket = ticketOpt.get();
        double fee = pricingService.calculateFee(ticket);
        Receipt receipt = receiptService.generateReceipt(ticket, fee);
        return receiptService.generateReceiptText(receipt, ticket);
    }

    // ─── Inner DTO ────────────────────────────────────────────────────────────

    public static class ExitResult {
        private final boolean success;
        private final UUID receiptId;
        private final double fee;
        private final String message;

        public ExitResult(boolean success, UUID receiptId, double fee, String message) {
            this.success = success;
            this.receiptId = receiptId;
            this.fee = fee;
            this.message = message;
        }

        public boolean isSuccess() { return success; }
        public UUID getReceiptId() { return receiptId; }
        public double getFee() { return fee; }
        public String getMessage() { return message; }

        @Override
        public String toString() {
            return "ExitResult{success=" + success
                    + ", receiptId=" + receiptId
                    + ", fee=₹" + fee
                    + ", message=" + message + "}";
        }
    }
}

File: Interview Problems/Parking Lot/Python Code/controller/exit_controller.py
import uuid
from typing import Optional
from service.ticket_service import TicketService
from service.pricing_service import PricingService
from service.payment_service import PaymentService
from service.receipt_service import ReceiptService
from service.slot_service import SlotService

class ExitResult:
    def __init__(self, success: bool, receipt_id: Optional[uuid.UUID], fee: float, message: str):
        self.success = success
        self.receipt_id = receipt_id
        self.fee = fee
        self.message = message

    def __repr__(self):
        return f"ExitResult(success={self.success}, receiptId={self.receipt_id}, fee=₹{self.fee:.2f}, message='{self.message}')"

class ExitController:
    def __init__(self, ticket_service: TicketService, pricing_service: PricingService,
                 payment_service: PaymentService, receipt_service: ReceiptService,
                 slot_service: SlotService):
        self.ticket_service = ticket_service
        self.pricing_service = pricing_service
        self.payment_service = payment_service
        self.receipt_service = receipt_service
        self.slot_service = slot_service
        print("[CONTROLLER] ExitController initialized")

    def exit_vehicle(self, ticket_id: uuid.UUID) -> ExitResult:
        print(f"[CONTROLLER] Vehicle exit request - Ticket: {ticket_id}")
        try:
            ticket = self.ticket_service.get_ticket(ticket_id)
            if not ticket:
                return ExitResult(False, None, 0.0, "Ticket not found")

            if not ticket.is_active:
                return ExitResult(False, None, 0.0, "Ticket is not active")

            fee = self.pricing_service.calculate_fee(ticket)
            print(f"[CONTROLLER] Fee calculated: ₹{fee}")

            payment_success = self.payment_service.process_payment_with_retry(ticket_id, fee, 3)
            if not payment_success:
                return ExitResult(False, None, fee, "Payment failed after retries")

            receipt = self.receipt_service.generate_receipt(ticket, fee)
            self.receipt_service.mark_receipt_as_paid(receipt)

            self.slot_service.release_slot(ticket.slot_id)
            self.ticket_service.deactivate_ticket(ticket_id)

            print(f"[CONTROLLER] Vehicle exit successful - Receipt: {receipt.id}")
            return ExitResult(True, receipt.id, fee, "Exit successful")

        except Exception as e:
            print(f"[CONTROLLER] Vehicle exit failed: {str(e)}")
            return ExitResult(False, None, 0.0, str(e))

    def generate_receipt_text(self, ticket_id: uuid.UUID) -> str:
        ticket = self.ticket_service.get_ticket(ticket_id)
        if not ticket:
            return f"Ticket not found: {ticket_id}"

        fee = self.pricing_service.calculate_fee(ticket)
        receipt = self.receipt_service.generate_receipt(ticket, fee)
        return self.receipt_service.generate_receipt_text(receipt, ticket)
